libcxx/include/flat_set

```diff
@@ -0,0 +1,913 @@
+#include <__utility/forward.h>
+#include <__utility/move.h>
+#include <__utility/pair.h>
+#include <__utility/sorted_equivalent.h>
+#include <__utility/sorted_unique.h>
+#include <vector>
+#include <version>
+
+// standard-mandated includes
+
+// [flat.set.syn]
+#include <compare>
+#include <initializer_list>
+
+#if !defined(_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER)
+#  pragma GCC system_header
+#endif
+
+#if _LIBCPP_STD_VER >= 23
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+template <class _FwdIt, class _Compare>
+_LIBCPP_HIDE_FROM_ABI _FwdIt
+__flatset_unique(_FwdIt __first, _FwdIt __last, reference_wrapper<const _Compare> __comp) {
+  if (__first == __last)
+    return __first;
+  _FwdIt __dest = __first;
+  for (++__first; __first != __last; ++__first) {
+    if (bool(__comp(*__dest, *__first))) {
+      ++__dest;
+      if (__dest != __first) {
+        *__dest = std::move(*__first);
+      }
+    }
+  }
+  ++__dest;
+  return __dest;
+}
+
+template <class _Key, class _Compare = less<_Key>,
+          class _KeyContainer = vector<_Key>>
+class flat_set {
+  static_assert(is_same_v<_Key, typename _KeyContainer::value_type>);
+  static_assert(__is_cpp17_random_access_iterator<typename _KeyContainer::iterator>::value);
+  static_assert(__is_cpp17_random_access_iterator<typename _KeyContainer::const_iterator>::value);
+
+public:
+  using key_type = _Key;
+  using value_type = _Key;
+  using key_compare = _Compare;
+  using value_compare = _Compare;
+  using reference = _Key&;
+  using const_reference = const _Key&;
+  using size_type = typename _KeyContainer::size_type;
+  using difference_type = typename _KeyContainer::difference_type;
+  using iterator = typename _KeyContainer::const_iterator;
+  using const_iterator = typename _KeyContainer::const_iterator;
+  using reverse_iterator = std::reverse_iterator<iterator>;
+  using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+  using container_type = _KeyContainer;
+
+  // [flat.set.cons]
+
+  flat_set()
+    noexcept(is_nothrow_default_constructible_v<container_type> && is_nothrow_default_constructible_v<key_compare>)
+    = default;
+
+  explicit flat_set(container_type __cont)
+    : __c_(std::move(__cont)) {
+    auto __begin = __c_.begin();
+    auto __end = __c_.end();
+    std::stable_sort(__begin, __end, __comp_ref());
+    __c_.erase(std::__flatset_unique(__begin, __end, __comp_ref()), __end);
+  }
+
+  template <class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(const container_type& __cont, const _Alloc& __a)
+    : __c_(std::__make_obj_using_allocator<container_type>(__a, __cont)) {
+    auto __begin = __c_.begin();
+    auto __end = __c_.end();
+    std::stable_sort(__begin, __end, __comp_ref());
+    __c_.erase(std::__flatset_unique(__begin, __end, __comp_ref()), __end);
+  }
+
+  template <class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(container_type&& __cont, const _Alloc& __a)
```

Comment:
Please `#if 0` this until the LWG issue is discussed and agreed upon, with a comment about the LWG issue.

Reply:
This is https://wg21.link/lwg3802 , and the fix is already done as a separate commit with "LWG3802" in the subject line.
The code for LWG3802 is unconditionally missing prior to the fix, and unconditionally present after the fix.
The tests for LWG3802 are `#if 0`'ed out prior to the fix, and unconditionally enabled after the fix.

If libc++ doesn't want to merge the fix for LWG3802 until [some-unspecified-point-in-the-future], that's fine; you can just omit that specific git commit. But I'm going to leave each LWG-issue-fixing commit in this actual PR branch for now, because that makes it easier to review the intended code/tests. (Otherwise you'll just ask over and over "Why is this part of the code wrong?" and the answer will always be "Because it doesn't become correct until after LWG issue such-and-such." It doesn't make sense to review wrong code.)